### PR TITLE
add support for constructors, nested classes, and chained methods

### PR DESCRIPTION
**Pull request by jonlouie**
### Issue #, if available:
N/A

### Description of changes:
* In `ClassDeclaration` model, added `BaseTypeOriginalDefinition` property to help identify generic base classes. 
  * Example: 
If a BaseType returns `System.Collections.Generic.List<MyConcreteClass>`, then BaseTypeOriginalDefinition returns `System.Collections.Generic.List<T>`
* Added support for identifying `ConstructorDeclaration` nodes
* Updated `GetNodes<T>()` method's traversal behavior to allow discovery of nested nodes of the same type (e.g. nested class declarations or chained method invocations)
* Created unit tests using new test project to support the above changes in business logic
* Updated existing unit test assertion to include Block nodes that are children of Constructor nodes

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/Analysis/Codelyzer.Analysis.CSharp/Handlers/ClassDeclarationHandler.cs [hunk 1]
```diff
@@ -21,6 +20,7 @@ public class ClassDeclarationHandler : UstNodeHandler
             if (classSymbol != null && classSymbol.BaseType != null)
             {
                 ClassDeclaration.BaseType = classSymbol.BaseType.ToString();
+                ClassDeclaration.BaseTypeOriginalDefinition = classSymbol.BaseType.OriginalDefinition.ToString();
```
Comment (nshiramshetti) [inline src/Analysis/Codelyzer.Analysis.CSharp/Handlers/ClassDeclarationHandler.cs:23]: Write a wrapper method to get the original definition. Please check for nulls inside.

File: src/Analysis/Codelyzer.Analysis.Model/ConstructorDeclaration.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿using System.Collections.Generic;
+using Newtonsoft.Json;
+ 
+namespace Codelyzer.Analysis.Model
+{
+    public class ConstructorDeclaration : UstNode
```
Comment (nshiramshetti) [inline src/Analysis/Codelyzer.Analysis.Model/ConstructorDeclaration.cs:6]: Can we use this class to extend MethodDeclaration - in addition, the method will have two more variables.

**Review by nshiramshetti — CHANGES_REQUESTED** [review]
Thanks for making the change. Some minor comments.  Also, please address warnings reported by the GitHub, like  

src/Analysis/Codelyzer.Analysis/AnalyzerCLI.cs#L62
The variable 'jsonData' is declared but never used

File: src/Analysis/Codelyzer.Analysis.CSharp/Handlers/ClassDeclarationHandler.cs (hunk 1 above)
Comment (jonlouie) [reply src/Analysis/Codelyzer.Analysis.CSharp/Handlers/ClassDeclarationHandler.cs]: Done

File: src/Analysis/Codelyzer.Analysis.Model/ConstructorDeclaration.cs (hunk 2 above)
Comment (jonlouie) [reply src/Analysis/Codelyzer.Analysis.Model/ConstructorDeclaration.cs]: I have extended the MethodDeclaration class with ConstructorDeclaration, but could you elaborate on the two method variables that are required?

**Comment by jonlouie** [comment]
@nshiramshetti Thank you for the feedback and helpful comments. I have addressed most issues, including the GitHub annotation checks. I just had the one question about the ConstructorDeclaration.

**Review by nshiramshetti — APPROVED** [review]
LGTM

**Pull request merged**